sympy/matrices/expressions/fourier.py

```diff
@@ -21,6 +23,32 @@ def _entry(self, i, j, **kwargs):
     def _eval_inverse(self):
         return IDFT(self.n)
 
+    def _eval_determinant(self):
+        eigs = self._eval_eigenvals()
+        if eigs:
+            tmp = 1
+            for e, cnt in eigs.items():
+                tmp *= e**cnt
+            return tmp
```

Comment:
It's generally more efficient to use a single call to `Mul`:
```
return Mul(*(e**c for e, c in eigs.items()))
```
It seems like this is always just some power of `I` though. Is there a formula for that?

Reply:
In the case of `DFT(0)` this will return an unsympified `int`.

It seems that `Matrix` also has that bug:
```julia
In [15]: type(Matrix([]).det())                                                                                                                
Out[15]: int
```

Reply:
For a formula let's consider the eigenvalues. For `DFT(n)` we have `m, r = divmod(n, 4)` and then if `r==0` then the eigenvalues with multiplicity are:
```
 1: m+1
-1: m
 I: m-1
-I: m
```
We can factor out the m part (assuming `m>0`) from the determinant so it becomes
```
(1**m * (-1)**m * I**n * -I**m)    *    (1 * I**-1)
```
Now we can strip out the `1 **m`. The `I**m` and `(-I)**m` combine to `(I*-I)**m = 1` and we're left with `(-1)**m`. Then the last part is `(1/I)` so the final result is just `(-1)**m / I` which gives:
```julia
In [28]: res.subs(n, 4)                                                                                                                        
Out[28]: ⅈ

In [29]: res.subs(n, 8)                                                                                                                        
Out[29]: -ⅈ
```
We can have similar formulas for each `r`.

Reply:
It should be possible to determine an equation based on modulo 8, but it wasn't any given in the paper I found (what I could see, scanned so hard to search). (Also, that would reduce the use `_eval_eigenvals`...)

```
d = [1, -1, I, I, -1, 1, -I, -I]
return d[(self.n -1) % 8]
```

Reply:
Maybe this can be simplified but for `n>0`:
```julia
In [25]: n = Symbol('n')                                                                                                                       

In [26]: det = ((-1)**(n//4) * (-1)**(n//2) * (-I)**((n-1)//2%2) )                                                                             

In [27]: det                                                                                                                                   
Out[27]: 
    ⎢n⎥     ⎢n⎥     ⎢n   1⎥      
    ⎢─⎥     ⎢─⎥     ⎢─ - ─⎥ mod 2
    ⎣4⎦     ⎣2⎦     ⎣2   2⎦      
(-1)   ⋅(-1)   ⋅(-ⅈ)  
```

Reply:
On the other hand, one can add the current `_eval_determinant`-code to a class higher up in the hierarchy and test if the `_eval_eigenvals` method exists, if it does one can compute the determinant from there. Which would be beneficial for a larger set of matrices. (Probably one would like that already in `Determinant.doit()`)

Reply:
That seems reasonable.